File: src/server/importer/importers/importer.service.ts
import { Inject, Injectable } from "@nestjs/common";
import { InjectModel } from "@nestjs/sequelize";

import { BankTransactionService } from "../../bank-transaction/bank-transaction.service";
import { BankTransactionBase } from "../../bank-transaction/models/bank-transaction.model";
import { Exception } from "../../core/errors/exception";
import { ensureOrThrow } from "../../core/errors/utils";
import { StatusReport } from "../models/status-report";
import { StatusReportRow } from "../models/status-report-row-import";

export const PARSERS_TOKEN = "PARSERS_TOKEN"

export interface ParserFactory {
    create(filePath: string) : Iterable<BankTransactionBase> | Promise<Iterable<BankTransactionBase>>
    fileRegex: string | RegExp
    key: string
}

@Injectable()
export class NewImportService {

    private parsersMap: Map<string, ParserFactory>

    constructor(
        @Inject(PARSERS_TOKEN) parsers: ParserFactory[],
        private readonly bankTransactionService: BankTransactionService,
        @InjectModel(StatusReport) private readonly statusReportModel: typeof StatusReport,
        @InjectModel(StatusReportRow) private readonly statusReportRowModel: typeof StatusReportRow
    ) {
        this.parsersMap = new Map(parsers.map(parser => [parser.key, parser]))
    }

    generateStatusTransaction(status: StatusReport, transaction: BankTransactionBase): StatusReportRow {
        return this.statusReportRowModel.build({
            ...transaction,
            reportId: status.id
        })
    }

    // eslint-disable-next-line max-statements, max-lines-per-function
    async import(groupOwnerId: number, key: string, fileName: string, filePath: string) {
        const status = this.statusReportModel.build({
            description: '',
            fileName,
            groupOwnerId,
            kind: key,
            status: 'OK',
        })
        try {
            await status.save()
            // Todo: validate we have this mapper
            const source = await ensureOrThrow(this.parsersMap.get(key), new Exception("E10003", 'Parser not found', {parserKey: key})).create(filePath)
            let discarting = true
            // eslint-disable-next-line init-declarations
            let previous : BankTransactionBase | undefined
            // eslint-disable-next-line init-declarations
            let previousState : StatusReportRow | undefined
            for await (const transaction of source){
                // Todo: validate transaction object
                const statusTransaction = this.generateStatusTransaction(status, transaction)
                if (await this.bankTransactionService.existsSimilar(groupOwnerId, key, transaction)){
                    if (discarting){
                        const msg = "repeated row, not inserted"

                        status.status = "WARN"
                        statusTransaction.message = msg
                        // eslint-disable-next-line max-depth
                        if (previousState){
                            previousState.message = msg
                            await previousState.save()
                            previousState = undefined
                        }
                        await statusTransaction.save()
                    } else {
                        previous = transaction
                        previousState = statusTransaction
                        discarting = true
                    }

                } else {
                    if (previous && previousState){
                        const record = await this.bankTransactionService.addTransaction({ kind: key, groupOwnerId, ...previous })
                        previousState.message = "Repeated row, but inserted"
                        previousState.transactionId = record.id
                        await previousState.save()
                        previous = undefined;
                        previousState = undefined;
                    }
                    discarting = false
                    const record = await this.bankTransactionService.addTransaction({ kind: key, groupOwnerId, ...transaction })
                    statusTransaction.transactionId = record.id
                    await statusTransaction.save()
                }

            }
        } catch (error) {
            const errorDict = Exception.getStdDict(error)
            status.description = errorDict.message
            status.stack = errorDict.stack
            status.context = errorDict.context ? JSON.stringify(errorDict.context) : undefined
            status.status = 'ERR'
            await status.save()
        }
        await status.save()
    }
}

File: swift-server/app/importers/importers/importer-service.swift
import Foundation
import Vapor

protocol ParserFactory {
	func create(filePath: String) -> AnySequence<BankTransaction>
	var fileRegex: String { get }
	var key: String { get }
}
/*
class NewImportService {
    private var parsersMap: [String: ParserFactory]

    private let bankTransactionService: BankTransactionService
    private let statusReportModel: StatusReport.Type
    private let statusReportRowModel: StatusReportRow.Type

    init(parsers: [ParserFactory], bankTransactionService: BankTransactionService, statusReportModel: StatusReport.Type, statusReportRowModel: StatusReportRow.Type) {
        self.parsersMap = Dictionary(uniqueKeysWithValues: parsers.map { ($0.key, $0) })
        self.bankTransactionService = bankTransactionService
        self.statusReportModel = statusReportModel
        self.statusReportRowModel = statusReportRowModel
    }

    func generateStatusTransaction(status: StatusReport, transaction: BankTransactionBase) -> StatusReportRow {
        return statusReportRowModel.init(reportId: status.id, transaction: transaction)
    }

    func `import`(groupOwnerId: Int, key: String, fileName: String, filePath: String) async throws {
        let status = statusReportModel.init(description: "", fileName: fileName, groupOwnerId: groupOwnerId, kind: key, status: "OK")

        do {
            try await status.save()

            guard let parser = parsersMap[key] else {
                throw Exception(.E10003, context: ["parserKey": key])
            }

            let source = parser.create(filePath: filePath)
            var discarting = true
            var previous: BankTransactionBase?
            var previousState: StatusReportRow?

            for try await transaction in source {
                let statusTransaction = generateStatusTransaction(status: status, transaction: transaction)

                if try await bankTransactionService.existsSimilar(groupOwnerId: groupOwnerId, key: key, transaction: transaction) {
                    if discarting {
                        let msg = "repeated row, not inserted"
                        status.status = "WARN"
                        statusTransaction.message = msg

                        if let prevState = previousState {
                            prevState.message = msg
                            try await prevState.save()
                            previousState = nil
                        }
                        try await statusTransaction.save()
                    } else {
                        previous = transaction
                        previousState = statusTransaction
                        discarting = true
                    }
                } else {
                    if let prev = previous, let prevState = previousState {
                        let record = try await bankTransactionService.addTransaction(kind: key, groupOwnerId: groupOwnerId, transaction: prev)
                        prevState.message = "Repeated row, but inserted"
                        prevState.transactionId = record.id
                        try await prevState.save()
                        previous = nil
                        previousState = nil
                    }
                    discarting = false
                    let record = try await bankTransactionService.addTransaction(kind: key, groupOwnerId: groupOwnerId, transaction: transaction)
                    statusTransaction.transactionId = record.id
                    try await statusTransaction.save()
                }
            }
        } catch {
            let errorDict = Exception.getStdDict(error)
            status.description = errorDict.message
            status.stack = errorDict.stack
            status.context = errorDict.context.map { JSONEncoder().encode($0) }
            status.status = "ERR"
            try await status.save()
        }

        try await status.save()
    }
}
*/
